ops/main.py

```diff
@@ -42,6 +43,38 @@
 logger = logging.getLogger()
 
 
+@dataclasses.dataclass(frozen=True)
+class _JujuContext:
+    # source: https://juju.is/docs/juju/charm-environment-variables
+    # the HookVars function: https://github.com/juju/juju/blob/3.6/worker/uniter/runner/context/context.go#L1398
+    # only allow accesing these vars because these are the currently used vars in main.py
+
+    juju_relation: str = ''
+    juju_relation_id: str = ''
```

Comment:
And I think this one should be `relation_id: Optional[int] = None`, that is, the parsed and converted version.